modules: trusted-firmware-a: introduce new module

SgrrZhf opened the pull request:
Add trusted-firmware-a as a module.

This PR is related to https://github.com/zephyrproject-rtos/zephyr/issues/46105

SgrrZhf committed:
boards: arm64: add fvp_base_revc_2xaemv8a_smp_ns board

fvp_base_revc_2xaemv8a_smp_ns board enable TFA and the zephyr code will
run in Normal world.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

stephanosio commented on west.yml:
```diff
@@ -228,6 +228,10 @@ manifest:
       path: modules/tee/tf-m/trusted-firmware-m
       groups:
         - tee
+    - name: trusted-firmware-a
```
Yes, you can temporarily point it to your own repository for testing.

microbuilder replied:
You could edit the `west.yml` file to point to your own repo, marking this as DNM while this is the case, with some variation of the instructions below, changing `revision: main` to an appropriate commit hash or branch:

```
For projects that have been setup using ``west``, you can add a local copy of
this module by adding the following sections to ``zephyr/west.yml``:

1. In the ``manifest/remotes`` section add:

.. code-block::

   remotes:
     - name: microbuilder
       url-base: https://github.com/microbuilder

2. In the ``manifest/projects`` section add:

.. code-block::

   - name: linaro_step
     remote: microbuilder
     path: modules/lib/step
     revision: main
```

This should at least let CI point to your repo until there is consensus to fork TF-A into Zephyr's github org.

SgrrZhf replied:
Thank you. I'll improve these.

stephanosio reviewed (commented):
There needs to be samples and tests that exercise this feature.

microbuilder commented:
@SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

The Corstone-1000 ecosystem FVP is configured with:

- An M0+ bootprocessor ([code available in TF-M](https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git/tree/platform/ext/target/arm/corstone1000))
- N x Cortex-A35 cores
- A secondary Cortex-M3 based on the [SSE-050 Subsystem](https://developer.arm.com/documentation/100918/0001/introduction/features-of-the-sse-050-subsystem)

We're (Linaro) currently exploring enabling the M3 on this board in Zephyr, using TF-M for the M0+ boot, and eventually we'd like to enable the A35 so that we have an option for A+M on Zephyr communicating with each other, so we can likely help with some of the technical burden here. @povergoing has my contact details if you want to discuss how we can avoid duplicating development efforts on this, if you agree Corstone-1000 FVP might be a good choice.

SgrrZhf commented:
> There needs to be samples and tests that exercise this feature.

Got that

povergoing commented:
> @SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

@microbuilder Great idea, no problem, Corstone-1000 is a very powerful platform. I think @SgrrZhf would like to create an issue to plan the future work, including enabling v8A SMP and sure A35 with TF-A and so on. We will contact you if there is some issue:)

povergoing commented:
> @SgrrZhf Would basing this off the Corstone-1000 make more sense, since there is an 'ecosystem' image available for that already for Linux and Windows, and it can be downloaded directly from Arm without license management, or authentication (making it very CI friendly: https://developer.arm.com/downloads/-/arm-ecosystem-fvps)?

Free to connect us if you need any help.
BTW, we also need the @stephanosio's help to make it into CI.

stephanosio commented:
> BTW, we also need the @stephanosio's help to make it into CI.

I can look into that as part of https://github.com/zephyrproject-rtos/zephyr/pull/45099.

SgrrZhf commented:
> There needs to be samples and tests that exercise this feature.

Could I create a directory 'samples/tfa-integration' and add a sample named 'hello_world'.  I will make more PR to enable more features such as PSCI later.

In this sample, I enable TFA. This sample will print messages like this and demonstrate that TF-A + zephyr is ok. 

```bash
huizha01@5d2d7be9157a:/workdir/zephyr/build-docker$ west build -t run
-- west build: running target run
[0/1] FVP_Base_RevC-2xAEMvA:
terminal_0: Listening for serial connection on port 5000
terminal_1: Listening for serial connection on port 5001
terminal_2: Listening for serial connection on port 5002
terminal_3: Listening for serial connection on port 5003

Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.flashloader0: FlashLoader: Loaded 168 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/fip.bin'

Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.secureflashloader: FlashLoader: Loaded 25 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/bl1.bin'
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL1: Built : 04:10:00, Jun  7 2022
NOTICE:  BL1: Booting BL2
NOTICE:  BL2: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL2: Built : 04:10:02, Jun  7 2022
NOTICE:  BL1: Booting BL31
NOTICE:  BL31: v2.7(release):v2.7.0-1-g371004ba9
NOTICE:  BL31: Built : 04:10:05, Jun  7 2022
*** Booting Zephyr OS build zephyr-v3.1.0-167-ge78f1ec86872  ***
Hello World! fvp_base_revc_2xaemv8a
```

SgrrZhf commented:
> > There needs to be samples and tests that exercise this feature.
> 
> Could I create a directory 'samples/tfa-integration' and add a sample named 'hello_world'. I will make more PR to enable more features such as PSCI later.
> 
> In this sample, I enable TFA. This sample will print messages like this and demonstrate that TF-A + zephyr is ok.
> 
> ```shell
> huizha01@5d2d7be9157a:/workdir/zephyr/build-docker$ west build -t run
> -- west build: running target run
> [0/1] FVP_Base_RevC-2xAEMvA:
> terminal_0: Listening for serial connection on port 5000
> terminal_1: Listening for serial connection on port 5001
> terminal_2: Listening for serial connection on port 5002
> terminal_3: Listening for serial connection on port 5003
> 
> Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.flashloader0: FlashLoader: Loaded 168 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/fip.bin'
> 
> Info: FVP_Base_RevC_2xAEMvA: FVP_Base_RevC_2xAEMvA.bp.secureflashloader: FlashLoader: Loaded 25 kB from file '/workdir/zephyr/build-docker/tfa/fvp/release/bl1.bin'
> NOTICE:  Booting Trusted Firmware
> NOTICE:  BL1: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL1: Built : 04:10:00, Jun  7 2022
> NOTICE:  BL1: Booting BL2
> NOTICE:  BL2: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL2: Built : 04:10:02, Jun  7 2022
> NOTICE:  BL1: Booting BL31
> NOTICE:  BL31: v2.7(release):v2.7.0-1-g371004ba9
> NOTICE:  BL31: Built : 04:10:05, Jun  7 2022
> *** Booting Zephyr OS build zephyr-v3.1.0-167-ge78f1ec86872  ***
> Hello World! fvp_base_revc_2xaemv8a
> ```

Any suggestions @stephanosio

stephanosio commented:
> Any suggestions @stephanosio

@SgrrZhf The sample(s) should be more comprehensive than just "hello world." Take a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration for example.

@carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

microbuilder commented:
> > Any suggestions @stephanosio
> 
> @SgrrZhf The sample(s) should be more comprehensive than just "hello world." Take a look at https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration for example.
> 
> @carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

`tfm_ipc` is probably the simplest example, and you should demonstrate a basic call from NS->S->NS: https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/tfm_integration/tfm_ipc

carlocaione commented:
> @carlocaione and @microbuilder probably have some idea as to what sample(s) should be there.

Well, having SMP working (thus PSCI) is also a good prof that TF-A is working correctly.

A long time ago I also added a PSCI test in `./tests/arch/arm64/arm64_psci` that should be good enough IMO.

povergoing commented:
Agreed with both, but shell we let @SgrrZhf pick only one of two options as the sample for this PR and then enable another one in the next PR? @microbuilder @stephanosio @carlocaione

SgrrZhf committed:
boards: arm64: fvp_base_revc_2xaemv8a: add more cpu nodes

fvp_base_revc_2xaemv8a supports up to 2 clusters and 8 cores, each
cluster has 4 cores.

Due to zephyr supports up to 4 cores, for now, I add another 3 cores,
they all belong to cluster0.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
boards: arm64: fvp_base_revc_2xaemv8a: change fvp parameters

- 'cluster0.NUM_CORES' can be configurable.
- Disable 'cache_state_modelled'

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
arch: arm64: init VMPIDR_EL2 in z_arm64_el2_init

VMPIDR_EL2 is assigned the value returned by EL2 reads of MPIDR_EL1

MPIDR_EL1 is the register holding the Multiprocessor ID which is to
identify different cores. Because of the virtualization requirements
for AArch64, MPIDR_EL1 should be virtualized (the different virtualized
cores can run on the same physical core). Thus the value of MPIDR_EL1
should be switched when the VM is switched. Setting the VMPIDR_EL2 is
the way to change the value returned by EL1 reads of MPIDR_EL1. Even
without virtualization, we still need to set VMPIDR_EL2 during booting
at EL2 or EL3. Otherwise, all cores' IDs are zero at the EL1 stage
which will break the SMP system.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
drivers: intc_gicv3: fix gic_rdists[cpu] points to a wrong address

In old way, gic_rdists[cpu] is calculated via MPIDR_TO_CORE(), but in
real hardware, MPIDR_TO_CORE() isn't a value increment from 0 one by
one, and that will lead gic_rdists[cpu] to point to a wrong address.

GICv3 provides the register GICR_TYPER[1] and it has a field named
Affinity_Value. This field can help to determine where gic_rdists[cpu]
should point.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
arch: arm64: fix the wrong way to send ipi interrupt

On GICv3, when we send an IPI interrupt, aff3, aff2 and aff1 should
be assigned a value corespond to a PE for which interrupt will be
generated. target_list only corresponds to aff0.

On real hardware, aff3, aff2, aff1 and aff0 should be treated as a
whole to determine a PE.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
tests: arm_smc_call: introduce arm smc call test

Add a simple test case to test the arm SMC call

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
modules: tfa: integrate trusted-firmware-a

This commit integrates trusted-firmware-a as a zephyr module

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
cmake: armfvp: permit run zephyr together with tf-a

As the TF-A was integrated into zephyr as a module. This adds necessary
CMake scripts to load BL1 and FIP binary built from TF-A.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf committed:
boards: fvp_baser_aemv8r: fix wrong reg size

Some reg size in fvp-aemv8r.dtsi isn't correct. This commit is used to
fix it.

- gic, the correct reg size is 0x10000 and 0x200000.
- uart0-3, the correct reg size is 0x10000.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -463,12 +463,55 @@ static void gicv3_dist_init(void)
+	for (mem_addr_t rdist_addr = GIC_RDIST_BASE;
+		rdist_addr < GIC_RDIST_BASE + GIC_RDIST_SIZE;
+		rdist_addr += 0x20000) {
+		uint64_t val = sys_read64(rdist_addr + GICR_TYPER);
```
CI reports _implicit declaration of function 'sys_read64'_

SgrrZhf committed:
arch: aarch32: cortex_a_r: introduce sys_read64

sys_read64 is useful when operating gicv3 register.

Signed-off-by: Huifeng Zhang <Huifeng.Zhang@arm.com>

SgrrZhf commented:
Hi, @carlocaione @stephanosio  @microbuilder

I'm finished.
Maybe I should put those fix patches to another PR to reduce the current PR's size? 
What do you think about it?

carlocaione commented:
I'm fine with this but IIUC given that you are modifying the `west.yml` file adding a new remote this cannot be merged as is.

SgrrZhf commented:
> I'm fine with this but IIUC given that you are modifying the `west.yml` file adding a new remote this cannot be merged as is.

Hi @carlocaione 
I added a temporary remote for testing, and the result is that all tests passed.
I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?

carlocaione commented:
> Hi @carlocaione I added a temporary remote for testing, and the result is that all tests passed. I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?

@stephanosio @nashif who should fork TF-A? Given that I assume @SgrrZhf is going to maintain it, is that something that he can do himself?

SgrrZhf commented:
Yes, @povergoing and I volunteer to maintain this(TF-A).

povergoing commented:
> > Hi @carlocaione I added a temporary remote for testing, and the result is that all tests passed. I wonder when could TF-A be forked into Zephyr's github org. Should I remove the temporary remote from `west.yml` now?
> 
> @stephanosio @nashif who should fork TF-A? Given that I assume @SgrrZhf is going to maintain it, is that something that he can do himself?

It would be much better that if there is something we could do it ahead, haha :)

stephanosio commented:
https://github.com/zephyrproject-rtos/trusted-firmware-a has been created.

stephanosio commented:
We need to add `West project: trusted-firmware-m` entry in the MAINTAINERS file.

SgrrZhf commented:
> We need to add `West project: trusted-firmware-a` entry in the MAINTAINERS file.
Got it.

povergoing commented on boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_smp_ns_defconfig:
```diff
@@ -0,0 +1,41 @@
+CONFIG_SERIAL=y
+
+# Enable serial port
+CONFIG_UART_PL011=y
```
Would you like to remove some CONFIGs that are not needed already, like `CONFIG_UART_PL011`?

SgrrZhf replied:
Yep

povergoing commented on boards/arm64/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a.dts:
```diff
@@ -32,6 +37,24 @@
+			reg = <0x200>;
+		};
+
+		 cpu@300 {
```
nit: a space before `cpu@300`?

SgrrZhf replied:
Ah, I see.

carlocaione commented on MAINTAINERS.yml:
```diff
@@ -2583,6 +2583,16 @@ West:
 
+"West project: trusted-firmware-a":
+  status: maintained
+  maintainers:
+    - povergoing
+    - sgrrzhf
```
Can I be added as collaborator? I will be quiet in a corner and I won't bother you, I promise :)

povergoing replied:
Hahaha, do u want to be one of the maintainers?

stephanosio replied:
In general, there should be no more than one maintainer.

carlocaione replied:
nah, you guys are the experts here, collaborator is fine so that I can be notified of PRs and code changes :) thanks!

SgrrZhf replied:
> In general, there should be no more than one maintainer.

I saw there are some area that has more than one maintainers, Could we have two maintainers here?

stephanosio replied:
This is not a particularly broad area, so one maintainer would be preferred; but, no strong opinion from me.

SgrrZhf replied:
Got it.

carlocaione commented on soc/arm64/arm/fvp_aemv8a/Kconfig.soc:
```diff
@@ -11,3 +11,10 @@ config SOC_FVP_BASE_REVC_2XAEMV8A
 
 endchoice
+
+config SOC_FVP_AEMV8A_EL2_INIT
+	bool "ARM FVP AEMv8A EL2 Initialization"
+	default n
+	help
+	  Initialize MPIDR_EL1 in z_arm64_el2_plat_init. It's necessary
+	  when zephyr starts at EL2 on multiprocessor systems.
```
Is this specific for FVP? If not this should be moved to the core code.

SgrrZhf replied:
Right, I'll move them.

carlocaione commented on tests/arch/arm64/arm64_svc_call/src/main.c:
```diff
@@ -0,0 +1,44 @@
+				  struct arm_smccc_res *);
+
+#ifdef CONFIG_SVC_CALL_USE_HVC
+	svc_call_method_t svc_call = arm_smccc_hvc;
```
nit: shouldn't this be called `smc_call_method_t` ?

SgrrZhf replied:
That's right.

stephanosio commented on arch/arm64/core/Kconfig:
```diff
@@ -193,6 +193,15 @@ config EXCEPTION_DEBUG
 
+config SET_VMPIDR_EL2
+	bool "Set VMPIDR_EL2 at EL2 stage"
+	default n
```
`default n` is redundant and should not be specified.

Also this config should be prefixed `ARM_` or `ARM64_` since this is specific to this arch.

stephanosio commented on boards/arm64/fvp_base_revc_2xaemv8a/board.cmake:
```diff
@@ -29,4 +29,16 @@ set(ARMFVP_FLAGS
+    set(FVP_SECURE_FLASH_FILE ${TFA_BINARY_DIR}/fvp/release/bl1.bin)
+    set(FVP_FLASH_FILE ${TFA_BINARY_DIR}/fvp/release/fip.bin)
+  endif()
+endif()
```
no new line at the end of file

SgrrZhf replied:
I didn't see any CMake file that need an extra new line at the end of the file. Do we really need it?

stephanosio commented on tests/arch/arm64/arm64_smc_call/Kconfig:
```diff
@@ -0,0 +1,8 @@
+
+config SMC_CALL_USE_HVC
+	bool "Set smc call method to hvc"
+	default n
```
same deal here

stephanosio commented on tests/arch/arm64/arm64_smc_call/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+tests:
+  arch.arm64.smc_call.smc:
+    arch_allow: arm
+    tags: arm drivers userspace
+    filter: CONFIG_BUILD_WITH_TFA
+  arch.arm64.smc_call.hvc:
+    arch_allow: arm
+    tags: arm drivers userspace
+    filter: CONFIG_BUILD_WITH_TFA
+    extra_configs:
+      - CONFIG_SMC_CALL_USE_HVC=y
```
These tests are being skipped by every platform (not run at all) because `BUILD_WITH_TFA` is not selected by any platforms.

SgrrZhf replied:
Got it.

stephanosio commented on tests/arch/arm64/arm64_smc_call/testcase.yaml:
```diff
@@ -0,0 +1,11 @@
+tests:
+  arch.arm64.smc_call.smc:
+    arch_allow: arm
+    tags: arm drivers userspace
+    filter: CONFIG_BUILD_WITH_TFA
+  arch.arm64.smc_call.hvc:
+    arch_allow: arm
+    tags: arm drivers userspace
+    filter: CONFIG_BUILD_WITH_TFA
+    extra_configs:
+      - CONFIG_SMC_CALL_USE_HVC=y
```
These tests are being skipped by every platform (not run at all) because `BUILD_WITH_TFA` is not selected by any platforms.

microbuilder commented on arch/arm64/core/Kconfig:
```diff
@@ -193,6 +193,15 @@ config EXCEPTION_DEBUG
+	bool "Set VMPIDR_EL2 at EL2 stage"
+	default n
+	help
+	  The purpose of VMPIDR_EL2 is that holds the value of the
+	  Virtualization Multiprocessor ID and this is the value returned
+	  by EL1 reads of MPIDR_EL1.
+	  Normally VMPIDR_EL2 is configurated at EL2 stage.
```
Didn't want to block merging for this, but since there are already some other change requests (thanks @stephanosio  for catching those), this should probably be updated for clarity:

```suggestion
	  VMPIDR_EL2 holds the value of the Virtualization Multiprocessor
	  ID. This is the value returned by EL1 reads of MPIDR_EL1.
	  Normally VMPIDR_EL2 is configured at the EL2 stage.
```

I'm not sure I follow the last line, though, since `SET_VMPIDR_EL2` defaults to `n`,
implying that this isn't 'normally' being set at the EL2 stage. (My experience is more
with Cortex-M, though, so I'm not as familiar with the registers here as I perhaps
should be to comment. The naming and comments aren't helping me, though.)

microbuilder replied:
@carlocaione Can you perhaps clarify or propose a clearer explanation of what this config setting is intended to do?

SgrrZhf replied:
That's right, and thanks for clarifying, I'll refine it.

povergoing commented:
Hi @stephanosio, would you like to take another look?
